Linux/sgx/test_app/enclave/tests/evp_smx.c

```diff
@@ -0,0 +1,617 @@
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ */
+
+#include <stdarg.h>
+#include <stdio.h> /* vsnprintf */
+#include <string.h>
+
+#include "openssl/bio.h"
+#include "openssl/evp.h"
+#include "openssl/ec.h"
+#include "openssl/pem.h" /* openssl library */
+
+#ifndef SAFE_FREE
+#define SAFE_FREE(ptr, size) do {if (NULL != (ptr)) {memset_s(ptr, size, 0, size); free(ptr); (ptr)=NULL;}} while(0);
+#endif
+
+// default sm2_user_id defined by openssl
+unsigned char sm2_user_id[] = "1234567812345678";
+unsigned int sm2_user_id_len = sizeof(sm2_user_id)-1;
+
+// create key pair including private key and public key
+int create_key_pair_sm2(EC_GROUP* ec_group, char** private_key, char** public_key)
+{
+    int ret = 0;
+    EC_KEY *ec_key = NULL;
+    BIO *pri_bio = NULL, *pub_bio = NULL;
+    size_t pri_len = 0, pub_len = 0;
+
+    do {
+        // 1. Create a new EC key
+        ec_key = EC_KEY_new();
+        if (ec_key == NULL) {
+            printf("Error: fail to create a new EC key\n");
+            ret = -1;
+            break;
+        }
+
+        // 2. Set the new EC key's curve
+        if (EC_KEY_set_group(ec_key, ec_group) != 1) {
+            printf("Error: fail to set the new EC key's curve\n");
+            ret = -2;
+            break;
+        }
+
+        // 3. Generate key pair based on the curve
+        if (!EC_KEY_generate_key(ec_key)) {
+            printf("Error: fail to generate key pair based on the curve\n");
+            ret = -3;
+            break;
+        }
+
+        // 4. Generate SM2 private key based on the curve
+        pri_bio = BIO_new(BIO_s_mem());
+        if (pri_bio == NULL) {
+            printf("Error: fail to create a BIO for private key\n");
+            ret = -4;
+            break;			
+        }
+        if (!PEM_write_bio_ECPrivateKey(pri_bio, ec_key, NULL, NULL, 0, NULL, NULL)) {
+            printf("Error: fail to write private key from ec_key to the BIO\n");
+            ret = -5;
+            break;			
+        }
+        pri_len = BIO_pending(pri_bio);
+        if (pri_len == 0) {
+            printf("Error: fail to get size of the BIO for private key\n");
+            ret = -6;
+            break;				
+        }
+        *private_key = (char*)malloc(pri_len);
+        if (BIO_read(pri_bio, *private_key, pri_len) <= 0) {
+            printf("Error: fail to read private key from the BIO\n");
+            SAFE_FREE(*private_key, sizeof(*private_key));
```

Comment:
sizeof(*private_key) should not be correct. Use pri_len instead. 

And You should check the ret value and free all the keys if ret !=0